TINF-376: Badgesystem

JulianRetzbach opened the pull request:
Ich habe die Definition des Badgesystems abgeschlossen, bin aber noch etwas unzufrieden mit den Level.

JulianRetzbach committed:
Badgesystem

Definition eines Badgesystems

jstsddg commented on api/v2/openapi.yaml:
```diff
@@ -428,6 +470,55 @@ components:
+        - date
+        - uccupancyReportId
+      properties:
+        Type:
```
Die einzelnen `properties` müssen in `lowerCamelCase` geschrieben werden :) Also das haben wir überall sonst auch so und das ist auch der `json`-Style wenn ich das richtig in Erinnerung habe. :D

JulianRetzbach replied:
hab ich angepasst

jstsddg commented on api/v2/openapi.yaml:
```diff
@@ -428,6 +470,55 @@ components:
+      description: Badge of a specific user
+      type: object
+      nullable: false
+      required:
```
Fehlt hier nicht noch der `type`?

JulianRetzbach replied:
Wo genau?

jstsddg commented on api/v2/openapi.yaml:
```diff
@@ -428,6 +470,55 @@ components:
+        Type:
+          $ref: '#/components/schemas/Badge'
+        UserId:
+          type: integer
```
`user` haben eine `UUID` als `id`, also müsste es
```
type: string
pattern: "^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$"
```
sein. `format` wird dann nicht mehr benötigt.

jstsddg commented on api/v2/openapi.yaml:
```diff
@@ -236,6 +236,48 @@ paths:
+          content:
+            application/json:
+              schema:
+                type: array
```
Also bekommt man bei dem Endpunkt alle Badges zurück die durch die letzte `Occupancy` hinzugekommen sind?

JulianRetzbach replied:
Anforderung war den neusten Badge zu bekommen, falls durch den letzten Report einer entstanden ist. Deshalb steht bei jedem Badge auch aus welchem Report er hervorgegangen ist.

jstsddg commented on api/v2/openapi.yaml:
```diff
@@ -428,6 +470,55 @@ components:
             $ref: '#/components/schemas/Location'
+    Badge:
+      title: Type
+          description: Badges for certain actions
+          type: string
+          nullable: true
+          enum:
+            - level1
+            - level2
+            - level3
+            - level4
+            - level4
+            - level5
+            - level6
+            - level7
+            - level8
+            - level9
+            - level10
+            - shortNewRateing
+            - shortNewShopRateing
+            - shortTwoDifferentShops
+            - shortShopNeverRated
+            - middleOneRateingPerWeek
+            - middleThreeRateingsPerWeek
+            - middleDailyRatingOneWeek
+            - middleOneWeekEachDayNewShopRated
+            - longFourWeeksSameShopRated
+            - longOneMonthDailyActive
```
Ab `description` ist hier alles `4 spaces` zu weit eingerückt, das muss parallel mit `title: Type` sein.

jstsddg commented on api/v2/openapi.yaml:
```diff
@@ -428,6 +470,55 @@ components:
           items:
             $ref: '#/components/schemas/Location'
+    Badge:
+      title: Type
```
Wenn das Schema `Badge` heißt, sollte dann nicht auch der Anzeigename (`title`) `Badge` statt `Type` sein? Oder zumindest `Badge Type`?

JulianRetzbach replied:
Hba ich jetzt passend geändert. Bei LocationType steht allerdings auch:
LocationType:
      title: Type

jstsddg commented on api/v2/openapi.yaml:
```diff
@@ -236,6 +236,48 @@ paths:
+        '401':
+          description: "Unauthorized: Not logged in"
+
+  '/users/self/newbadge{id}':
```
Ich glaube das `{id}` soll da nicht hin ;) Was hältst du von `/users/self/badges/newest`? Da dieser Endpunkt ja auch zu dem `badges` Endpunkt gehört?

jstsddg replied:
Dann sollte man die `operationId` auch zu `listNewestBadges` oder sowas umbennen, damit das alles ungefähr gleich bleibt, oder was meinst du?

JulianRetzbach replied:
Macht Sinn. Mir ist auch grad aufgefallen, dass man die id des Reports gar nicht braucht, weil man den User sowieso hat

jstsddg commented on api/v2/openapi.yaml:
```diff
@@ -236,6 +236,48 @@ paths:
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/Badge'
```
Meintest du hier vielleicht `UserBadges`? Sonst verstehe ich nicht warum du das `schema` dafür angelegt hast.

JulianRetzbach replied:
Hab ich später geändert und vergessen zu übernehmen

jstsddg reviewed (changes requested):
Ich habe mir das Konzept wie `Badges` überhaupt im gesamten funktionieren noch nicht angeschaut. Ich schaue mir die Tickets dafür nochmal an :)

JulianRetzbach committed:
Update nach Review

Einige Fehler wurden behoben

JulianRetzbach committed:
Update openapi.yaml

Einrückungsfehler behoben

jstsddg committed:
Merge branch 'master' into TINF-376-Badge-System

JulianRetzbach committed:
Update openapi.yaml

Änderungen an Badgesystem nach Absprache mit dem Frontend.
Enum Stage noch mit Platzhaltern gefüllt.

JulianRetzbach committed:
Merge branch 'TINF-376-Badge-System' of https://github.com/SAKPaaS/documentation into TINF-376-Badge-System

jstsddg commented on api/v2/openapi.yaml:
```diff
@@ -496,6 +557,111 @@ components:
+        level:
+          type: integer
+          example: 5
+        xpTotal:
+          type: integer
+          example: 150
+        xpNeeded:
+          type: integer
+          example: 25
+        xpCurrent:
+          type: integer
+          example: 10
```
Das würde ich selbst nochmal in ein Objekt fassen, weil ja alles nur mit `xp` zu tun hat, also:
```json
{
"stage": "novice",
"level": 1,
"xp": {
    "total": 150,
    "needed": 25,
    "current": 10
}}
```

bzw:
```yaml
properties:
  stage:
    $ref: '#/components/schemas/Stage'
  level:
    type: integer
  xp:
    type: object
    properties:
      total:
        type: integer
      needed:
        type: integer
      current:
        type: integer
```

JulianRetzbach replied:
Sieht so wirklich besser aus. Werd ich übernehmen.

jstsddg commented on api/v2/openapi.yaml:
```diff
@@ -496,6 +557,111 @@ components:
+        - badgeAndCounter
+        - countAllBadges
+      properties:
+        BadgeAndCounter:
```
Properties werden in `lowerCamelCase` geschrieben. Und `badges` wäre hier ausreichend, sonst sind die Namen irgendwann überladen.

JulianRetzbach replied:
ok

jstsddg commented on api/v2/openapi.yaml:
```diff
@@ -496,6 +557,111 @@ components:
+            type: array
+            items:
+              $ref: '#/components/schemas/BadgeAndCounter'
+        NumberOfBadges:
```
Auch hier finde ich den Namen etwas überladen, ich würde einfach `count` vorschlagen.

jstsddg commented on api/v2/openapi.yaml:
```diff
@@ -496,6 +557,111 @@ components:
+        reason:
+          type: string
+          example: occupancy
+    AllBadges:
```
`AllBadges` klingt so als wären das alle existierenden Badges, wie wäre es mit `UserBadges`?

jstsddg commented on api/v2/openapi.yaml:
```diff
@@ -496,6 +557,111 @@ components:
+        - week_special
+        - trusted_shop
+        - trusted_user
+    NewBadge:
```
`BadgeUpdate` wäre hier meiner Meinung nach zukunfts sicherer, falls `Badges` noch andere Funktionen bekommen.

jstsddg commented on api/v2/openapi.yaml:
```diff
@@ -496,6 +557,111 @@ components:
           maxItems: 100
           items:
             $ref: '#/components/schemas/Location'
+    Badge:
```
Zu einer Badge gehört doch immer die Anzahl, oder? Dann kann man doch direkt ein Objekt mit `type` und `count` draus machen? Dann braucht man auch nicht diesen sperrigen `BadgeAndCount` Datentyp.

JulianRetzbach replied:
das Objekt mit type und count gibt es doch in ähnlicher Form. Heißt jetzt aber wie du es angeregt hast Badges.

jstsddg commented on api/v2/openapi.yaml:
```diff
@@ -496,6 +557,111 @@ components:
+        counter:
+          type: integer
+          example: 1
+    LevelAndXp:
```
Reicht nicht auch hier einfach nur `Level`, bzw. `UserLevel` da man für die `Level` ja eh `xp` braucht?

jstsddg commented on api/v2/openapi.yaml:
```diff
@@ -496,6 +557,111 @@ components:
+      type: string
+      nullable: true
+      enum:
+        - stage1
+        - stage2
+        - stage3
```
Das bearbeitet gerade @DavidSchmidt00 noch, darauf müssen wir warten.

JulianRetzbach replied:
hat ja noch Zeit

jstsddg commented on api/v2/openapi.yaml:
```diff
@@ -496,6 +557,111 @@ components:
+        - xpCurrent
+      properties:
+        stage:
+          type: 
+            $ref: '#/components/schemas/Stage'        
```
Du kannst entwerder einen `type` setzen oder `$ref` verwenden, aber nicht beides. Also:
```yaml
stage:
  type: object
```
oder
```yaml
stage:
  $ref: '#/components/schemas/Stage'
```

JulianRetzbach committed:
Update openapi.yaml

Änderungen nach Review durch Joost. Änderungen an Stages noch ausstehend.

JulianRetzbach committed:
Update Stages

stages updated from https://docs.google.com/document/d/1O4fJnQsdkLveY8TcFBdvYPaIxlTgynsua7djrwG7FKY/edit

luisafuss commented on api/v2/openapi.yaml:
```diff
@@ -277,6 +277,67 @@ paths:
+    get:
+      operationId: levelAndXp
+      description: |-
+        provides the the level, stage, total xp, xp needed for the next level and current xp for the logged in user
```
```suggestion
        provides the level, stage, total xp, xp needed for the next level and current xp in this level for the logged in user
```

luisafuss commented on api/v2/openapi.yaml:
```diff
@@ -496,6 +557,110 @@ components:
+          example: 5
+    Badges:
+      title: Badges
+      description: shows how often a user earnd the same badge
```
```suggestion
      description: shows how often a user earned the same badge
```

luisafuss reviewed (changes requested):
Kenne mich damit nicht wirklich aus, aber das was wir besprochen hatten, scheint so drin zu sein. Hab nur kleine sprachliche Anmerkungen

DavidSchmidt00 commented on api/v2/openapi.yaml:
```diff
@@ -496,6 +557,110 @@ components:
+      title: BadgeUpdate
+      description: latest Badges of a specific user
+      type: object
+      nullable: false
```
müsste das nicht nullable sein, wenn kein neues Badge dazu kam?

DavidSchmidt00 replied:
Oder zeigt das allgemein nur das letzte Badge an was dazu kam?
Weil dann würde ja die Funktionalität fehlen, dass man iwo abfragen kann, ob beim letzten report ein badge dazu kam

DavidSchmidt00 replied:
@MrKraboom @JulianRetzbach

jstsddg replied:
@DavidSchmidt00 Habe dein Problem gefixt, indem im `BadgeUpdate` dann `reason: no_update` steht

DavidSchmidt00 replied:
perfect

JulianRetzbach committed:
Update api/v2/openapi.yaml

Änderung in der Formulierung

Co-authored-by: luisafuss <54306427+luisafuss@users.noreply.github.com>

JulianRetzbach committed:
Update api/v2/openapi.yaml

Änderung in der Formulierung

Co-authored-by: luisafuss <54306427+luisafuss@users.noreply.github.com>

JulianRetzbach commented:
@DavidSchmidt00 passt das Ganze so wie es aktuell umgesetzt ist?

MrKraboom commented:
@JulianRetzbach es gibt noch eins zwei Fehler, bin aber noch nicht dazu gekommen das Review vollständig zu machen. Ich kann auch anbieten die zu fixen und dann anzunehmen?

jstsddg committed:
Small Fixes and naming changes

jstsddg reviewed (approved):
LGTM

luisafuss reviewed (approved):
Sieht gut aus. Scheint alles da zu sein
